File: Go/Lesson_006_binary_tree_preorder_traversal.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 006 -- Binary Tree Preorder Traversal
// Category   : Trees
// Difficulty : Easy
// Study Plan : Day 3
// =============================================================
//
// QUESTION:
//   Given the root of a binary tree, return the preorder (Root, Left, Right)
//   traversal of its nodes' values.
//
//   Example:
//     Input : root = [1,null,2,3]
//     Output: [1, 2, 3]
// =============================================================

package main

import "fmt"

type TreeNode struct { Val int; Left, Right *TreeNode }

func preorderTraversal(root *TreeNode) []int {
    if root == nil { return []int{} }
    out := []int{}; st := []*TreeNode{root}
    for len(st) > 0 {
        n := st[len(st)-1]; st = st[:len(st)-1]
        out = append(out, n.Val)
        if n.Right != nil { st = append(st, n.Right) }
        if n.Left  != nil { st = append(st, n.Left) }
    }
    return out
}

func main() {
    r := &TreeNode{Val:1, Right:&TreeNode{Val:2, Left:&TreeNode{Val:3}}}
    fmt.Println(preorderTraversal(r))
}

File: Java/Lesson006_BinaryTreePreorderTraversal.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 006 -- Binary Tree Preorder Traversal
// Category   : Trees
// Difficulty : Easy
// Study Plan : Day 3
// =============================================================
//
// QUESTION:
//   Given the root of a binary tree, return the preorder (Root, Left, Right)
//   traversal of its nodes' values.
//
//   Example:
//     Input : root = [1,null,2,3]
//     Output: [1, 2, 3]
// =============================================================

import java.util.*;

public class Lesson006_BinaryTreePreorderTraversal {
    static class TreeNode { int val; TreeNode left, right; TreeNode(int v){val=v;} }

    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> out = new ArrayList<>();
        if (root == null) return out;
        Deque<TreeNode> st = new ArrayDeque<>(); st.push(root);
        while (!st.isEmpty()) {
            TreeNode n = st.pop(); out.add(n.val);
            if (n.right != null) st.push(n.right);
            if (n.left  != null) st.push(n.left);
        }
        return out;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1); root.right = new TreeNode(2); root.right.left = new TreeNode(3);
        System.out.println(new Lesson006_BinaryTreePreorderTraversal().preorderTraversal(root));
    }
}
